cpp/include/cudf/detail/iterator.cuh

```diff
@@ -0,0 +1,207 @@
+ * @brief value accessor of column with null bitmask
+ * A unary functor returns scalar value at `id`.
+ * `operator() (cudf::size_type id)` computes `element` and valid flag at `id`
+ * This functor is only allowed for nullable columns.
+ *
+ * the return value for element `i` will return `column[i]`
+ * if it is valid, or `null_replacement` if it is null.
+ *
+ * @throws `cudf::logic_error` if the column is not nullable.
+ *
+ * @tparam Element The type of elements in the column
+ * -------------------------------------------------------------------------**/
+template <typename Element>
+struct column_device_view::null_replaced_value_accessor
+{
+  column_device_view const col;         ///< column view of column in device
+  Element const null_replacement{};     ///< value returned when element is null
+
+/** -------------------------------------------------------------------------*
+ * @brief constructor
+ * @param[in] _col column device view of cudf column
+ * @param[in] null_replacement The value to return for null elements
+ * -------------------------------------------------------------------------**/
+  null_replaced_value_accessor(column_device_view const& _col, Element null_val)
+    : col{_col}, null_replacement{null_val}
+  {
+    // verify valid is non-null, otherwise, is_valid() will crash
+    CUDF_EXPECTS(_col.nullable(), "Unexpected non-nullable column.");
+  }
+
+  CUDA_DEVICE_CALLABLE
+  Element operator()(cudf::size_type i) const {
+    return col.is_valid_nocheck(i) ?
+      col.element<Element>(i) : null_replacement;
+  }
+};
+
+/** -------------------------------------------------------------------------*
+ * @brief value accessor of column without null bitmask
+ * A unary functor returns scalar value at `id`.
+ * `operator() (cudf::size_type id)` computes `element`
+ * This functor is only allowed for non-nullable columns.
+ *
+ * the return value for element `i` will return `column[i]`
+ *
+ * @throws `cudf::logic_error` if the column is nullable.
+ *
+ * @tparam Element The type of elements in the column
+ * -------------------------------------------------------------------------**/
+
+template <typename Element>
+struct column_device_view::value_accessor
+{
+  column_device_view const col;         ///< column view of column in device
+
+/** -------------------------------------------------------------------------*
+ * @brief constructor
+ * @param[in] _col column device view of cudf column
+ * -------------------------------------------------------------------------**/
+  value_accessor(column_device_view const& _col)
+    : col{_col}
+  {
+    // verify column does not have nulls
+    CUDF_EXPECTS(!_col.has_nulls(), "Unexpected column with nulls.");
+  }
+
+  CUDA_DEVICE_CALLABLE
+  Element operator()(cudf::size_type i) const {
+    return col.element<Element>(i);
+  }
+};
+
+template <typename T>
+column_device_view::const_iterator<T> column_device_view::begin() const {
```

Comment:
I don't like how the implementation of a `column_device_view` member function is split off into another header. The implementation should be in `column_device_view.cuh`.

Reply:
`begin` and `end` moved to `column_device_view.cuh`. `value_accessor` is still in `iterator.cuh` because of type_dispatcher.hpp include. (otherwise column_device_view.cuh is crowded with too many includes). should I keep `value_accessor` in `iterator.cuh`?

Reply:
Why does `value_accessor` need `type_dispatcher.hpp`?

Reply:
It has type checking of column datatype with parameter type. `type_to_id` is defined in `type_dispatcher.hpp`
```cpp
    CUDF_EXPECTS(data_type(experimental::type_to_id<Element>()) == col.type(),
                 "the data type mismatch");
```

Reply:
moved.